KEP 4216: Add changes for alpha version under RuntimeClassInImageCriApi feature gate

kiashok opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Add changes for alpha release of Image pull per runtime class feature described in KEP 4216: https://github.com/kubernetes/enhancements/issues/4216
  - Adds feature gate "RuntimeClassInImageCriApi" and necessary related alpha changes

#### Which issue(s) this PR fixes:

Add changes for alpha release of Image pull per runtime class feature described in KEP 4216: https://github.com/kubernetes/enhancements/issues/4216
  - Adds feature gate "RuntimeClassInImageCriApi" and necessary related alpha changes
  
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
NONE 

```release-note
Added new feature gate called "RuntimeClassInImageCriApi" to address kubelet changes needed for KEP 4216.
Noteable changes:
1. Populate new RuntimeHandler field in CRI's ImageSpec struct during image pulls from container runtimes.
2. Pass runtimeHandler field in RemoveImage() call to container runtime in kubelet's image garbage collection
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


```docs

- [KEP]: https://github.com/kubernetes/enhancements/commit/5d2744011e205f6d759bac6ab316a085bec2a605

```

kiashok committed:
Imagepull per runtime class alpha release changes

This commit does the following:
1. Add RuntimeClassInImageCriApi feature gate
2. Extend pkg/kubelet/container Image struct
3. Adds runtimeHandler string in the following CRI calls
   i.   ImageStatus
   ii.  PullImageRequest
   iii.  RemoveImage

Signed-off-by: kiashok <kiashok@microsoft.com>

kiashok commented:
cc @jsturtevant @marosset @aravindhp @mrunalp @haircommander @mikebrow

aravindhp commented on pkg/features/kube_features.go:
```diff
@@ -699,6 +699,13 @@ const (
+	// alpha: v1.29
+	//
+	// Adds support to pull images based on the runtime class specified
+	RuntimeClassInImageCriApi featuregate.Feature = "RuntimeClassInImageCriApi"
```
Should this be added in isolation without the code where it is being used? We don't want a situation where the feature gate merges but rest of the code misses the release.

kannon92 replied:
Yea I would add this with the code you want to be guarded by the feature.

mikebrow replied:
yeah suggest draft the pr till get the code and any nec. test improvements..

kiashok replied:
I've updated the PR with the necessary changes under the feature gate. I am not sure adding tests would make sense yet as no changes have been implemented on the container runtime side (example in containerd) and we cannot validate through functional or e2e tests in k8s. I have validated it against a private draft version of the containerd changes on AKS cluster.

mikebrow replied:
see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/image_manager_test.go#L338

mikebrow replied:
also in the e2e node checks it would be ok to write the test bucket and add "skips" for the container runtime releases that are not ready yet.. can do manual testing if not available in any.. and the skips get removed when support avail and the gate is on

kiashok replied:
Hmm I am not sure we would be able to test anything at all meaningfully with the fakeruntimes. I feel it would make more sense to add these tests once some functionality is in container runtimes like containerd atleast. As of now all this is doing is to read runtimehandler field that is already in podspec and setting ImageSpec.Runtimehandler.

Any thoughts on this @aravindhp @jsturtevant @marosset @haircommander ?

kannon92 replied:
I think adding unit tests to cover the feature toggling logic would be very important.  You won't really cover those in e2e tests.  

I think you can add tests pretty easily to all these points.  You don't need to have a e2e test but it at least covers the case of the feature toggle off/on and logic makes sense.

kiashok replied:
@kannon92 https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4216-image-pull-per-runtime-class#can-the-feature-be-disabled-once-it-has-been-enabled-ie-can-we-roll-back-the-enablement turning feature off and testing needs manual steps as mentioned in the KEP. I am not sure that can be tested? We could test to ensure ListImages returns with the runtimeHandler of the image set etc

kannon92 replied:
Yea for E2E I understand that.  I still think you could test `ListImages` with feature toggle.  Same for things like convert.go, image_manager, etc.  You actually should try to test those components with unit as it would be difficult to cover those areas with e2e tests.  E2E tests can be used to verify the full implementation including container runtime but there is no reason why that should stop you from writing unit tests with the fake CRI.

kiashok replied:
@mikebrow updated PR to include test. Please let me know if this looks good. Thanks.

kiashok replied:
Updated PR with all unit tests for functionality added under the new feature gate

jsturtevant commented:
/priority important-soon
/sig windows
/sig node
/milestone v1.29

kiashok commented:
/retest

haircommander commented:
personally, I am more of a fan of writing the base struct and overriding with an if statement, rather than defining the struct in two places:
```

			im.imageRecords[image.ID] = &imageRecord{
				firstDetected: detectTime,
			}
			if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
				im.imageRecords[image.ID].runtimeHandlerUsedToPullImage = image.Spec.RuntimeHandler
			}
```
and so fourth. It's a stylistic thing, and if you prefer the way it is, keep it. but if it's no different to you, I'd prefer the other way :)

kiashok commented:
<img width="973" alt="image" src="https://github.com/kubernetes/kubernetes/assets/99994218/1c834e5a-0629-4ddb-b003-f93cd8a933b1">
The tests are failing with an unrelated error. Should this error be fixed in a separate PR? 

cc @haircommander

kiashok commented:
> personally, I am more of a fan of writing the base struct and overriding with an if statement, rather than defining the struct in two places:
> 
> ```
> 
> 			im.imageRecords[image.ID] = &imageRecord{
> 				firstDetected: detectTime,
> 			}
> 			if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
> 				im.imageRecords[image.ID].runtimeHandlerUsedToPullImage = image.Spec.RuntimeHandler
> 			}
> ```
> 
> and so fourth. It's a stylistic thing, and if you prefer the way it is, keep it. but if it's no different to you, I'd prefer the other way :)

I'll make this change where applicable and push new changes.

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -105,6 +108,16 @@ func (m *kubeGenericRuntimeManager) ListImages(ctx context.Context) ([]kubeconta
+		// Therefore, validate that image returned from ListImages() call above has a valid
+		// runtimehandler field set; that is, it should not be ""
+		if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+			if img.Spec != nil && img.Spec.RuntimeHandler == "" {
```
you should be able to show that in the mock test linked above ^

kiashok replied:
@mikebrow added test. Please let me know if this looks good. Thanks.

kiashok replied:
The only intention here is to have a uniform way to identify as an image as its name/ID + runtime class used to pull the image. Does k8s allow us to always have the feature gate to support this feature? If yes, we can enable the feature gate for runtimes that implement this feature. If not, I am not sure how clean the implementation of this feature would be to have some images identified as name and empty "" runtime handler and in some other cases images to be identified as name and a non-empty runtime handler. 
For example is this behavior acceptable - 
 1. Lets say with CRI-O runtime, some user specifies a runtime handler in the deployment yaml, then the EnsureImageExists() call is going to pass the runtime handler down to the CRI-O container runtime which may or may not choose to use this field. Likewise, when kubelet GC calls ListImages() from the container runtime, then CRI-O responds but does  not set the runtime handler field and GC continues to function as it is today.
 2. Lets say the container runtime is containerd which implements image pull per runtime class changes on its side completely (that is, an image on containerd will be identified as a tuple of imageName/imageId, runtimeHandler and if an empty runtimeHandler is passed down to it via CRI, then it uses the default runtime handler for that platform.
 Likewise, when kubelet GC calls ListImages() from containerd, containerd returns the list of images with the imageName and the runtime handler used to pull that image and then kubelet  GC uses both these fields to decide which image it wants to delete.

If the above behavior is acceptable from k8s side and seems like a decent solution to have given its flexibility, we could do as @haircommander is suggesting. Thoughts @mikebrow @marosset @jsturtevant @aravindhp ?

kiashok replied:
Had a discussion with @mikebrow over a call now. Here's what might be a better approach:

Had a discussion with @mikebrow over a call now. Here's what might be a better approach:

1. On image pull and garbage collection:

- if runtime handler is specified in the yaml, EnsureImageExists() will pass on the imageName and runtimeHandler field to the container runtime and the container runtime
can choose to use this field or not.
- In kubelet image GC, the following is an important data structure used:

type realImageGCManager struct {
	// Container runtime
	runtime container.Runtime

	// Records of images and their use.
	imageRecords     map[string]*imageRecord  >>>>>>>>>>>>>>> map of ImageRecord struct that is indexed by ImageID
	imageRecordsLock sync.Mutex

	// The image garbage collection policy in use.
	policy ImageGCPolicy

	// statsProvider provides stats used during image garbage collection.
	statsProvider StatsProvider

	// Recorder for Kubernetes events.
	recorder record.EventRecorder

	// Reference to this node.
	nodeRef *v1.ObjectReference

	// Track initialization
	initialized bool

	// imageCache is the cache of latest image list.
	imageCache imageCache

	// tracer for recording spans
	tracer trace.Tracer
}


the detectImages() function creates new imageRecord entries for new images after calling container runtime's ListImages() call. The index to realImageGCManager.imageRecords[] should now 
be a tuple of  (imageID, runtimeHandler) instead of just the imageID.

If the container runtime implements KEP 4216, then ListImages() call will return a list of imageID, runtimeHandler. If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216. If runtime != "", then great the container runtime implements KEP 4216 like containerd.


2. What if the user changes the definition of the runtime handler on container runtime side like containerd. 

Example:

(i) a runtime class could have orginially been

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-hyperv.guest_platform]
            Architecture = "amd64"
            OS = "windows"
            OSFeatures = []
            OSVersion = "10.0.20348"
            Variant = ""


(ii) Then it was changed to the following:

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process.guest_platform]
            Architecture = "amd64"
            OS = "windows"
            OSFeatures = []
            OSVersion = "10.0.17763"
            Variant = ""

When runtimeclass information is changed, this would need a restart of containerd but the old images pulled with (i) will still be loaded back on containerd restart and kubelet will think the image exists but the runhcs-wcow-process.guest_platform information has changed. Therefore, we need to have a TODO in containerd code and also k8s documentation for this KEP
that if the runtimeclass's information changes, we need to flush containerd and restart both containerd and kubelet.


cc @jsturtevant @marosset @aravindhp @haircommander

jsturtevant replied:
> Does k8s allow us to always have the feature gate to support this feature?

The typical pattern is for it to be turned on by default in `Beta` then locked to on once in `GA`.  With the feature gate being dropped after that

I was under the impression that empty string would equate to default runtime handler and there for not throw an error here based on
https://github.com/kubernetes/kubernetes/blob/6fed03ea91627185afd4976474a396f17689f0d7/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L787

> Lets say with CRI-O runtime, some user specifies a runtime handler in the deployment yaml, then the EnsureImageExists() call is going to pass the runtime handler down to the CRI-O container runtime which may or may not choose to use this field.

This sounds like it could be an edge case where the pulled image might not be the expected version.  Should we log a warning here in this case?  In this case I think user would experience the same errors they do with out this feature?

jsturtevant replied:
> If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216

You still have the situation described with CRI-O. Is the KEP 4216 something runtimes must implement? if so a failure here is warranted otherwise a warning makes sense to me.

kiashok replied:
> I was under the impression that empty string would equate to default runtime handler and there for not throw an error here based on

>> This is true on the container runtime side when we implement the support for this feature. If runtimehandler returned in ListImages() is "", kubelet is not aware of default runtime

kiashok replied:
> > If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216
> 
> You still have the situation described with CRI-O. Is the KEP 4216 something runtimes must implement? if so a failure here is warranted otherwise a warning makes sense to me.

I don't think KEP 4216 should be a must implement for all runtimes. if no runtimeHandler is returned in the list images response from the container runtime, then the index tuple for this image should be (imageId, "") in the imageRecords map of kubelet GC

aravindhp replied:
>that if the runtimeclass's information changes, we need to flush containerd and restart both containerd and kubelet.

Will this require the node to be cordoned and drained? Not sure who the "we" here is that will restart both the services.

kiashok replied:
The admin decides if runtime class definition was changed and flushes/remove images from containerd  and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?

jsturtevant replied:
> If runtimehandler returned in ListImages() is "", kubelet is not aware of default runtime

We should make this clear in the docs as an expectation.  

> I don't think KEP 4216 should be a must implement for all runtimes.

this should not error here then, I still think there is an edge case where the user specifies the runtimeclass and it is sent to the runtime and it doesn't handle it properly and it should be logged as a warning.

kiashok replied:
@jsturtevant Agreed. I haven't made changes to reflect these in the PR yet as I was looking for feedback/arrive at a consensus for the proposal I made earlier about desired behavior. I'll work on these now and push need changes.

aravindhp replied:
>The admin decides if runtime class definition was changed and flushes/remove images from containerd and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?

It was not immediately obvious to me that admin intervention was required. Not sure if this is a great UX but I will defer to SIG-node here.

kiashok replied:
> > The admin decides if runtime class definition was changed and flushes/remove images from containerd and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?
> 
> It was not immediately obvious to me that admin intervention was required. Not sure if this is a great UX but I will defer to SIG-node here.

Not sure that there is a better way to do this. If not, the existing images can be allowed to be loaded and used by containerd and eventually starting containers could fail and then user would have to flush containerd of the older images. The latter might be a painful debugging experience. There it might be better to flush containerd and restart just like in toggling the feature gate for this KEP would need manual intervention. 
Alternatively, we could add an option/annotation/config on containerd side which when set in the toml causes containerd to purge and delete the images using the edited runtime class rather than loading them. Any thoughts on this last option @mikebrow @dmcgowan?

haircommander replied:
we had to make a similar concession in the [cgroup driver reported by runtime KEP](https://github.com/kubernetes/enhancements/issues/4033). Basically, it's up to the container runtime/admin to verify this information isn't changing while there are containers/images that already exist. Kubelet has no way to verify it mid run (or it would be inefficent to do so). Explicitly declaring this is a requirement of the CRI is good here, and having a mechanism in the CRI to report errors to the admin would be good to (but out of scope here)

haircommander replied:
this invariant https://github.com/kubernetes/kubernetes/pull/118770/files#diff-7a3c2723dedc6b44cf183fc52f84cc4015ffbd65a998a65c3959a953cb63c945R139-R141 it's declared in the CRI definition

kiashok replied:
> this invariant https://github.com/kubernetes/kubernetes/pull/118770/files#diff-7a3c2723dedc6b44cf183fc52f84cc4015ffbd65a998a65c3959a953cb63c945R139-R141 it's declared in the CRI definition

This is great! 
"The Kubelet will not re-request the RuntimeConfiguration after startup, and CRI implementations should avoid updating them without a full node reboot." 
-- Node reboot might still not fix containerd loading previous images in containerd's store. I discussed this in the containerd community meeting this morning and we are fine with the existing invariant which is to ask admins to purge if runtimeclass name is the same but some of its fields like platform versioning and snapshotters have changed. 

cc @mikebrow @dmcgowan

kiashok replied:
@haircommander @jsturtevant @aravindhp @mikebrow pushed new changes based on the discussion. Updated the tests as necessary. 
Please let me know if it looks ok :) Thanks a lot!

kannon92 commented:
> <img alt="image" width="973" src="https://user-images.githubusercontent.com/99994218/277740674-1c834e5a-0629-4ddb-b003-f93cd8a933b1.png"> The tests are failing with an unrelated error. Should this error be fixed in a separate PR?
> cc @haircommander

I think you gotta fix this in this PR.  That script is usaully pretty concret in detecting issues

You can reproduce this with the verify scripts.  I think `make verify` or `hack/verify*` should reproduce this.  

It looks like there are some formatting issues

kiashok commented:
> > <img alt="image" width="973" src="https://user-images.githubusercontent.com/99994218/277740674-1c834e5a-0629-4ddb-b003-f93cd8a933b1.png"> The tests are failing with an unrelated error. Should this error be fixed in a separate PR?
> > cc @haircommander
> 
> I think you gotta fix this in this PR. That script is usaully pretty concret in detecting issues
> 
> You can reproduce this with the verify scripts. I think `make verify` or `hack/verify*` should reproduce this.
> 
> It looks like there are some formatting issues

I'll take a look. Its about camel cases. 
Just curious - how did these changes even make it through in the first place? This PR isn't touching that file at all.

kannon92 commented:
> > > <img alt="image" width="973" src="https://user-images.githubusercontent.com/99994218/277740674-1c834e5a-0629-4ddb-b003-f93cd8a933b1.png"> The tests are failing with an unrelated error. Should this error be fixed in a separate PR?
> > > cc @haircommander
> > 
> > 
> > I think you gotta fix this in this PR. That script is usaully pretty concret in detecting issues
> > You can reproduce this with the verify scripts. I think `make verify` or `hack/verify*` should reproduce this.
> > It looks like there are some formatting issues
> 
> I'll take a look. Its about camel cases. Just curious - how did these changes even make it through in the first place? This PR isn't touching that file at all.

I am not sure but it looks like it also saying that the code that you added was not formatted with go-fmt.  You could try `hack/update-gofmt.sh` to run that and commit.  I'm not sure of the error you are seeing but maybe that it is a red herring.

moficodes commented:
Hello @kiashok ! 👋🏼
Bug triage for the 1.29 release cycle is here!
This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 2 weeks from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.

As the issue is tagged for 1.29, is it still planned for that release?

bart0sh commented:
/triage accepted

kannon92 commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -509,3 +523,62 @@ func TestEvalCRIPullErr(t *testing.T) {
+
+	useSerializedEnv := true
+	t.Run(c.testName, func(t *testing.T) {
+		featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.RuntimeClassInImageCriAPI, true)()
```
I think you need a defer here.  At the end of this test, it will set the toggle back to previous setting.  No need fo rline 551.  

I also expect this will fail the linter as we check that defer is required for these feature gates.

kannon92 replied:
I usually see cases where one uses a test table with feature off and feature on.  

I would expect to see a case where feature is off and another test case where feature is on.

kiashok commented:
> Hello @kiashok ! 👋🏼 Bug triage for the 1.29 release cycle is here! This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 1 week from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.
> 
> As the issue is tagged for 1.29, is it still planned for that release?

@moficodes which issue are you referring to? There was a CRI change PR that made it into the main branch last week and this PR was only submitted yesterday. 
But yes, we are targeting this for 1.29

kiashok commented:
> > > > <img alt="image" width="973" src="https://user-images.githubusercontent.com/99994218/277740674-1c834e5a-0629-4ddb-b003-f93cd8a933b1.png"> The tests are failing with an unrelated error. Should this error be fixed in a separate PR?
> > > > cc @haircommander
> > > 
> > > 
> > > I think you gotta fix this in this PR. That script is usaully pretty concret in detecting issues
> > > You can reproduce this with the verify scripts. I think `make verify` or `hack/verify*` should reproduce this.
> > > It looks like there are some formatting issues
> > 
> > 
> > I'll take a look. Its about camel cases. Just curious - how did these changes even make it through in the first place? This PR isn't touching that file at all.
> 
> I am not sure but it looks like it also saying that the code that you added was not formatted with go-fmt. You could try `hack/update-gofmt.sh` to run that and commit. I'm not sure of the error you are seeing but maybe that it is a red herring.

updated the PR after running hack/update-gofmt.sh

kannon92 commented:
> > Hello @kiashok ! 👋🏼 Bug triage for the 1.29 release cycle is here! This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 1 week from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.
> > As the issue is tagged for 1.29, is it still planned for that release?
> 
> @moficodes which issue are you referring to? There was a CRI change PR that made it into the main branch last week and this PR was only submitted yesterday. But yes, we are targeting this for 1.29

I believe @moficodes is referring to this issue: https://github.com/kubernetes/enhancements/issues/4216.

Can you add this PR to the alpha requirements?

mikebrow commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -248,8 +252,15 @@ func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.
-				firstDetected: detectTime,
+			if !utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+				im.imageRecords[image.ID] = &imageRecord{
+					firstDetected: detectTime,
```
note: this works when the differentiator for the container runtime - runtime class config is a platform/version change that results in a different digest value (sha256) because at least one bit in the binary for that platform/version is different, but it does not work, I don't believe, when the differentiator for the runtime class is the snapshotter. 

As discussed in the KEP, one would need to create a tuple (imageName/imageDigest + runtimeClass) here to allow for identification if the image is ready based on this entry in the list.

mikebrow replied:
resolution may be as simple as return the differentiating fields in the image pull response/list apis.. and concatenate them in the ID as a parseable string... thus allowing the ID+runtime handler platform/version+snapshotter/fs to be used as the index..

kiashok replied:
@mikebrow I am not sure I understand the question/concern here. Could you please clarify?
As mentioned in KEP 4216, we are not aiming to support snapshotter per runtime class at all with this kep. Here were are just going to be identifying image as name/ID and runtime class

mikebrow replied:
per discussion we agreed a TODO may be fine for extending to the "edited" runtime handler config situation, as you said, we can just require a restart to refresh ... and thus be able to trust two runtime handler configs will have two different names or "" and a non default name..

kiashok replied:
@mikebrow  summarized the discussion we had here - https://github.com/kubernetes/kubernetes/pull/121456#discussion_r1372104416

kiashok replied:
Addressed this in the new changes that I just pushed. Please let me know if it looks ok. Thanks!

mikebrow reviewed (commented):
I think we need a tuple func to identify image records by image.ID(digest)+runtimeHandler.. at a minimum, if not runtimeHandler.platform + runtimeHandler.storage (storage = snapshotter and/or fs location currently defaulted for the runtime handler)

Though if the runtime config is changed the runtimeHandler contents may be switched up...

runtime handler = runc 
runc config set to windows + version, snapshotter "default" 
user edits config to runc handler is now windows + (version++), snapshotter "forADifferentFS i.e. block"
user restarts containerd... 
we might need some way to identify parts of the runtime config that have meaning with respect to the image.ID and the assumption that ref is ready for use for the container w/pod specified runtime handler.

jsturtevant commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -105,6 +107,17 @@ func (m *kubeGenericRuntimeManager) ListImages(ctx context.Context) ([]kubeconta
+		// field is empty and log a warning message.
+		if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+			if img.Spec == nil || (img.Spec != nil && img.Spec.RuntimeHandler == "") {
+				klog.V(3).InfoS("WARNING: RuntimeHandler is empty", "ImageID", img.Id)
```
Is this worth `v(2)`?

jsturtevant replied:
also no sure on the style here.  Are there other places where `WARNING` is used?

kiashok replied:
klog.Warning() is giving a lint error for using unstructured logging. When I checked other places, they were using a klog.V().InfoS() with WARNING in the log message

jsturtevant replied:
yup `klog.warning()` isn't recommended: https://github.com/kubernetes/community/blob/01cc099c45882a5e83c796f2b8a787901bfa93dd/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions  

Sounds good on WARNING with infos.  wasn't sure what the precedent was

kiashok commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -90,7 +97,11 @@ type realImageGCManager struct {
+	// If RuntimeClassInImageCriAPI feature gate is enabled, imageRecords
+	// are identified by a tuple of (imageId,runtimeHandler) that is passed
+	// from ListImages() call. If no runtimehandler is specified in response
+	// to ListImages() by the container runtime, "" is used.
```
yeah, it would be imageid- . This is the most generic case when container runtimes have not implemented this kep.

kiashok replied:
updated it to your suggestion to use just imageID when the runtimeHandler field is ""

kiashok commented on pkg/kubelet/kuberuntime/convert_test.go:
```diff
@@ -150,3 +153,142 @@ func TestConvertToRuntimeAPIImageSpec(t *testing.T) {
+	}
+
+	for _, test := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.RuntimeClassInImageCriAPI, true)()
```
Adding it to the test before this function

kiashok commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -375,6 +375,7 @@ func TestPullAndListImageWithPodAnnotations(t *testing.T) {
 
 		image := images[0]
 		assert.Equal(t, "missing_image:latest", image.ID, "Image ID")
+		assert.Equal(t, "", image.Spec.RuntimeHandler, "image.Spec.RuntimeHandler not empty", "ImageID", image.ID)
```
Commit was missed accidentally. Bringing it back in

kiashok commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -182,14 +182,17 @@ func TestDetectImagesWithNewImage(t *testing.T) {
 	require.True(t, ok)
 	assert.Equal(detectedTime, newContainer.firstDetected)
 	assert.Equal(zero, noContainer.lastUsed)
+	assert.Equal("", noContainer.runtimeHandlerUsedToPullImage)
```
Image GC is using the list of pods, its containers and looking into kubecontainer.Container.ImageID to build list of used images. Unable to find a way in fakeruntime to populate the newly added ImageRuntimeHandler field in kubecontainer.Container. Without that, the imageTuple calculation will not go through as expected for the case when feature gate is set

kiashok replied:
found a way to do this. Updating the tests

moficodes commented:
Hello @kiashok  ! 👋🏼
Bug triage for the 1.29 release cycle is here!
This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is this week. We want to make sure that every PR has a chance to be merged on time.

As the issue is tagged for 1.29, is it still planned for that release?

jsturtevant commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -457,10 +538,23 @@ func (im *realImageGCManager) imagesInEvictionOrder(ctx context.Context, freeTim
+				imageRecord: *record,
+			})
+		} else {
+			imageTuples := strings.Split(image, ",")
```
suggestion: Maybe make this a function like `getImageTuple` to make it  clear that is what is happening here.

haircommander replied:
not a huge deal but you didn't use the helper here

kiashok replied:
@haircommander I haven't pushed changes to address James's review comments still. Are you also suggesting the same as James is or am I missing something?

haircommander replied:
I thought that's what https://github.com/kubernetes/kubernetes/pull/121456/files#diff-887859ee44bc4f73de8b1a5e92feab5bda8b12fa5f452859d46037d6eaefc3adR549 was

kiashok replied:
@haircommander Hit the push command by mistake but hit ctrl-C. Didn't realize that it had already gone through. Aplogies for the confusion! The PR is updated now with all comments addressed. Thanks!

kiashok commented:
> Hello @kiashok ! 👋🏼 Bug triage for the 1.29 release cycle is here! This issue hasn't been updated for a long time, so I wanted to check what the status is. The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is this week. We want to make sure that every PR has a chance to be merged on time.
> 
> As the issue is tagged for 1.29, is it still planned for that release?

Hi @moficodes , yes we are still targeting 1.29 . Have addressed review comments and waiting for feedback. Updated the github issue with the latest status

haircommander commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -222,7 +236,20 @@ func (im *realImageGCManager) GetImageList() ([]container.Image, error) {
+// from ListImages() call. If no runtimehandler is specified in response
+// to ListImages() by the container runtime, only imageID will be will
+// be returned.
+func getImageTuple(imageID, runtimeHandler string) string {
```
this is a personal nit: but I always prefer defining functions below the last caller so you can read the file iteratively down.

haircommander commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -423,19 +483,40 @@ func (im *realImageGCManager) freeSpace(ctx context.Context, bytesToFree int64,
+	} else {
+		// if runtimehandler is empty when RuntimeClassInImageCriAPI is enabled, log info message
+		// and call RemoveImage() without setting image's RuntimeHandler
+		if image.imageRecord.runtimeHandlerUsedToPullImage == "" {
+			klog.V(5).InfoS("RuntimeHandler is empty", "imageID", image.id)
+			err = im.runtime.RemoveImage(ctx, container.ImageSpec{Image: image.id})
+		} else {
+			klog.InfoS("Removing image to free bytes", "imageID", image.id, "size", image.size, "runtimeHandler", image.runtimeHandlerUsedToPullImage)
+			err = im.runtime.RemoveImage(ctx, container.ImageSpec{Image: image.id, RuntimeHandler: image.runtimeHandlerUsedToPullImage})
```
I would prefer this is compressed to 
```
klog.InfoS("Removing image to free bytes", "imageID", image.id, "size", image.size, "runtimeHandler", image.runtimeHandlerUsedToPullImage)
err = im.runtime.RemoveImage(ctx, container.ImageSpec{Image: image.id, RuntimeHandler: image.runtimeHandlerUsedToPullImage})
```
we don't need the if because RuntimeHandler will already be "" if we don't specify it (like on like 496 now) and the message "RuntimeHandler is empty" is the same as `runtimeHandler: ""` in a structured log

haircommander commented:
one more round of nits @kiashok then LGTM thanks for your continual updates!

haircommander commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -423,19 +470,37 @@ func (im *realImageGCManager) freeSpace(ctx context.Context, bytesToFree int64,
+			klog.V(5).InfoS("RuntimeHandler is empty", "imageID", image.id)
+		}
+		klog.InfoS("Removing image to free bytes", "imageID", image.id, "size", image.size, "runtimeHandler", image.runtimeHandlerUsedToPullImage)
+		err = im.runtime.RemoveImage(ctx, container.ImageSpec{Image: image.id, RuntimeHandler: image.runtimeHandlerUsedToPullImage})
```
I agree with this throughout.  I think `runtimeHandler: ""` in the structured log is acceptible to reduce the code

haircommander commented:
/lgtm

thanks @kiashok

haircommander commented:
still 
/lgtm

marosset commented:
@kiashok 
> ERROR: pkg/kubelet/images/image_gc_manager.go:533:6: ST1003: func getImageIdFromTuple should be getImageIDFromTuple (stylecheck)
ERROR: func getImageIdFromTuple(image string) string {

marosset commented:
/lgtm

Thanks @kiashok!

jsturtevant commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -457,15 +495,46 @@ func (im *realImageGCManager) imagesInEvictionOrder(ctx context.Context, freeTim
+			imageID := getImageIDFromTuple(image)
+			// Ensure imageID is valid or else continue
+			if imageID == "" {
+				klog.V(2).InfoS("WARNING: ImageID is not valid, skipping", "ImageID", imageID)
```
what's the impact of skipping?

jsturtevant replied:
Should this be a raised event?

jsturtevant commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -248,28 +269,36 @@ func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.
-		klog.V(5).InfoS("Adding image ID to currentImages", "imageID", image.ID)
-		currentImages.Insert(image.ID)
+		imageKey := image.ID
+		if !isRuntimeClassInImageCriAPIEnabled {
+			klog.V(5).InfoS("Adding image ID to currentImages", "imageID", imageKey)
+		} else {
+			imageKey = getImageTuple(image.ID, image.Spec.RuntimeHandler)
+			klog.V(5).InfoS("Adding image ID with runtime class to currentImages", "imageKey", imageKey, "runtimeHandler", image.Spec.RuntimeHandler)
+		}
```
lower in this file the log messages just include runtimeHandler and leave it "" if not enabled.  Since we are doing that below should we follow the same pattern here?

```suggestion
		if isRuntimeClassInImageCriAPIEnabled {
			imageKey = getImageTuple(image.ID, image.Spec.RuntimeHandler)
		}
		
		klog.V(5).InfoS("Adding image ID to currentImages", "imageID", imageKey, "runtimeHandler", image.Spec.RuntimeHandler)
```

kiashok replied:
![image](https://github.com/kubernetes/kubernetes/assets/99994218/adbe3a18-faf4-49c3-945b-655b158cfa0d)

Wanted to explicitly call out the difference between imageID and imageKey when featuregate is enabled atleast for alpha phase. That's why I left it unchanged

jsturtevant commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -237,8 +252,14 @@ func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.
 		for _, container := range pod.Containers {
-			klog.V(5).InfoS("Container uses image", "pod", klog.KRef(pod.Namespace, pod.Name), "containerName", container.Name, "containerImage", container.Image, "imageID", container.ImageID)
-			imagesInUse.Insert(container.ImageID)
+			if !isRuntimeClassInImageCriAPIEnabled {
+				klog.V(5).InfoS("Container uses image", "pod", klog.KRef(pod.Namespace, pod.Name), "containerName", container.Name, "containerImage", container.Image, "imageID", container.ImageID)
+				imagesInUse.Insert(container.ImageID)
+			} else {
+				imageKey := getImageTuple(container.ImageID, container.ImageRuntimeHandler)
+				klog.V(5).InfoS("Container uses image", "pod", klog.KRef(pod.Namespace, pod.Name), "containerName", container.Name, "containerImage", container.Image, "imageID", container.ImageID, "imageKey", imageKey)
+				imagesInUse.Insert(imageKey)
+			}
```
same logging comment here: 

```suggestion
                        imageKey := container.ImageID
			if isRuntimeClassInImageCriAPIEnabled {
				imageKey := getImageTuple(container.ImageID, container.ImageRuntimeHandler)
			}
			
			klog.V(5).InfoS("Container uses image", "pod", klog.KRef(pod.Namespace, pod.Name), "containerName", container.Name, "containerImage", container.Image, "imageID", container.ImageID, "imageKey", imageKey)
			imagesInUse.Insert(imageKey)
```

kiashok replied:
Wanted to explicitly call out the difference between imageID and imageKey. That is why I didn't change these two if-else blocks:
![image](https://github.com/kubernetes/kubernetes/assets/99994218/fc6fcbf8-98f9-492a-bdce-757e3c11daf4)

jsturtevant commented:
/lgtm

jsturtevant commented:
@kiashok can you add a release note?

kiashok commented:
> @kiashok can you add a release note?

Added one now

kiashok commented:
/test pull-kubernetes-unit

kiashok commented:
All the tests are passing

jsturtevant commented:
/assign @sergeykanzhelev @dchen1107  
for node feature-gate approval

kiashok commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

mikebrow reviewed (approved):
/lgtm
Thank you!

mrunalp commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -105,6 +107,17 @@ func (m *kubeGenericRuntimeManager) ListImages(ctx context.Context) ([]kubeconta
+		// field is empty and log a warning message.
+		if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+			if img.Spec == nil || (img.Spec != nil && img.Spec.RuntimeHandler == "") {
+				klog.V(2).InfoS("WARNING: RuntimeHandler is empty", "ImageID", img.Id)
```
What is it expected to be set for images that aren't pulled with a runtime handler? If that is going to be empty, we don't want to warn on every such image.

kiashok replied:
Container runtimes like containerd that intend to make changes for this KEP will always respond with a runtimehandler when EnsureImageExists() calls into ImageStatus on container runtime. for runtimes that implement the changes needed, even if runtimehandler is passed in as "", container runtime will see that and convert it to the default runtime handler for that platform. This will just be an info warning for container runtimes that do not make changes for this KEP which is fine. If this warning shows up for runtimes like containerd that implement this feature this would show a bug

mrunalp commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -141,6 +171,64 @@ func TestDetectImagesInitialDetect(t *testing.T) {
+				{
+					ID:      container.ContainerID{Type: "test", ID: fmt.Sprintf("container-%d", 1)},
+					ImageID: imageID(1),
+					// The image filed is not set to simulate a no-name image
```
nit: field

derekwaynecarr commented on pkg/kubelet/container/runtime.go:
```diff
@@ -282,6 +284,8 @@ type Container struct {
 	// The id of the image used by the container.
 	ImageID string
+	// Runtime handler used to pull the image if any.
+	ImageRuntimeHandler string
```
I get why `ImageSpec` was updated to be inclusive of `RuntimeHandler`, and how that is convenient when runtime manager `ListImages` is invoked (where there is no associated pod or container context).

What is not clear to me is why `Container` is extended to have `ImageRuntimeHandler` rather than `Pod`.  The `RuntimeClassName` is a pod level attribute, and it looks like every other call site where we work with `Container` we have access to the encapsulating `Pod` context.  Worried in future we could have a mismapped reference of some kind.  I do not think this is blocking on the PR, but it felt more natural to me to model that runtime handler against the pod since the API facing the users has the field at `Pod` scope.

kiashok replied:
![image](https://github.com/kubernetes/kubernetes/assets/99994218/c71ab1b7-e6c4-4c7e-aacd-2123daef9180)

I had to do this because of how image GC is reading imageID (its doing it from the kuberuntime.Container.ImageId) and the image is referred to by just the imageID in kuberuntime.Container which should not be the case once we move to identifying images as (imageID, runtimeHandler).

kiashok replied:
If we do not want to add runtimeHandler to kuberuntime.Container, I think we should get rid of the imageId too as it might cause issues in the future if we just refer to the image that a container is using by the imageID and not (imageId, runtimeHandler). We could have the same ImageID for different runtime handlers and it would be risky to just reference the image by ImageID only

derekwaynecarr replied:
@kiashok that was the code block that was tripping me up.  `ListImages` needs to include the runtime handler, totally agree.  If runtimeHandler was added to `Pod`, it would be returned in `runtime.GetPods` with that pod.

In the code block you capture above, would imagesInUse would just use the imageKey of `container.ImageID,pod.RuntimeHandler`, no?

derekwaynecarr replied:
reflecting a bit on the comment to remove `imageID`, i guess that is fair.  The issue is that is the ID according to the runtime, its just only identifying when now associated with a runtime handler... I think that is a separate issue we should think about how to cleanup, but for this particular instance, it feels like Pod is probably the right place to associate runtime handler information in kuberuntime.

kiashok replied:
Hi @derekwaynecarr , I think Container struct having the imageID is kinda expected since container runtimes also use imageID to identify which image the container is actively using now. This also seems to be expected from Container struct in CRI:

![image](https://github.com/kubernetes/kubernetes/assets/99994218/bccc0cb3-4190-41ce-a8a1-4332d2d97ec7)

It would be useful to know the imageID of the image that the container is currently using right? I am not sure we would want to get rid of it completely though as container runtimes might need to know what image a container needs and if its present or not right? I think now that we are changing how an image is identified, we would need to just have info about runtimeHandler as well. That's why I named the field as "ImageRuntimeHandler" with that comment that it is the runtimehandler used to pull the image. I still feel its important to identify what image is being used by container. Any thoughts?

jsturtevant replied:
> In the code block you capture above, would we not just insert into imagesInUse the tuple key composed of container.ImageID and pod.RuntimeHandler?

Initially makes sense to me


@kiashok  If I understand correctly this concern is there might be a mis match between what containerd knows about the image and what kubelet does? 

One thing I noticed is the cri `container` message shouldn't need `ImageRuntimeHandler` since it already has the `Image` reference in the `container` message and `ImageSpec` will have that information? https://github.com/kubernetes/kubernetes/blob/029452198566a41bc39d04a1ec5bad3f37621a1c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1195

Would that resolve the concern of a mis-match and not require that field added?

mikebrow replied:
from a container... should be able to find the runtime handler like so: 
pod := podSandboxMap[container.podSandboxID]   m.runtimeClassManager.LookupRuntimeHandler(pod.Spec.RuntimeClassName)

kiashok replied:
@jsturtevant , I was referring to kubecontainer.Container struct and not CRI container. CRI container struct already has reference to ImageSpec which is correct. kubecontainer.Container struct in pkg/kubelet/container/runtime.go has its own struct and saves info as needed:

![image](https://github.com/kubernetes/kubernetes/assets/99994218/3e4e73ba-e2da-4c3c-8e78-dab25a8cb5de)


I think the concern is we do not need to have imageRuntimeHandler in this struct.

I think it would be good to have runtimehandler here as they are using imageID as reference to the image. It might cause confusions as image is no longer just recognized by the imageID

kiashok replied:
@mikebrow I know we could go backwards and get it. My concern is that we should make image runtime handler explicit in structures that are using imageID to reference the image currently in use by the container

kiashok replied:
If container CRI has reference to ImageSpec which has a field for runtimeHandler, we should maintain uniformity and also have image runtime handler with kuberuntime.Container?

kiashok replied:
> @kiashok that was the code block that was tripping me up. `ListImages` needs to include the runtime handler, totally agree. If runtimeHandler was added to `Pod`, it would be returned in `runtime.GetPods` with that pod.
> 
> In the code block you capture above, would we not just insert into `imagesInUse` the tuple key composed of `container.ImageID` and `pod.RuntimeHandler`?

@derekwaynecarr , that code is changed in the current PR. I was just pointing out to what it is like in the current kubelet package.

for reference:

kiashok replied:
Since this is not a blocker , can this be added as a follow up in beta after further discussion?

derekwaynecarr replied:
I’m good with follow up.

mikebrow replied:
nod.. pls add a TODO for follow up revision

kiashok replied:
Discussed with @mikebrow  on slack. This will be TODO for beta and I'll add this to sig-node meeting agenda for the next meeting and have a wider discussion with folks before updating the repo with these changes. thanks!

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -125,6 +127,16 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+		runtimeHandler := ""
+		if pod.Spec.RuntimeClassName != nil && *pod.Spec.RuntimeClassName != "" {
+			runtimeHandler = *pod.Spec.RuntimeClassName
```
could you check here.. might have to look it up: m.runtimeClassManager.LookupRuntimeHandler(pod.Spec.RuntimeClassName)

kiashok replied:
@mikebrow I did figure out that we could go backwards and get the info. But my concern what rising from wanting to having uniformity with how images are being identified.

https://github.com/kubernetes/kubernetes/pull/121456#discussion_r1378092610

https://github.com/kubernetes/kubernetes/pull/121456#discussion_r1378094084

mikebrow replied:
unless the Spec.RuntimeClassName is the runtimeHandler.. I think you need to call LookupRuntimeHandler here

mikebrow replied:
change lgtm (passing the hander through Ensure)..

mikebrow commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -237,8 +252,14 @@ func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.
+				klog.V(5).InfoS("Container uses image", "pod", klog.KRef(pod.Namespace, pod.Name), "containerName", container.Name, "containerImage", container.Image, "imageID", container.ImageID)
+				imagesInUse.Insert(container.ImageID)
+			} else {
+				imageKey := getImageTuple(container.ImageID, container.ImageRuntimeHandler)
```
right here I think we might be able to get away with finding the handler by looking up the pod runtime class

kiashok replied:
Added response in previous comment thread: https://github.com/kubernetes/kubernetes/pull/121456#discussion_r1378100936

kiashok replied:
https://github.com/kubernetes/kubernetes/pull/121456#discussion_r1378166991 @mikebrow please let me know if this is ok

kiashok replied:
Discussed with @mikebrow on slack. This will be TODO for beta and I'll add this to sig-node meeting agenda for the next meeting and have a wider discussion with folks before updating the repo with these changes. thanks!

mikebrow replied:
track and change this to address scoping is pod in a follow up... this given the runtime manager where the runtime class manager is ... is not readily available to the image managers

mikebrow replied:
sgtm

kiashok commented:
![image](https://github.com/kubernetes/kubernetes/assets/99994218/fb8eaeff-4bda-4643-8e35-03f44185e247)

This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be? 

cc @kannon92

kannon92 commented:
> ![image](https://user-images.githubusercontent.com/99994218/279509072-fb8eaeff-4bda-4643-8e35-03f44185e247.png)
> 
> This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be?
> 
> cc @kannon92

I notice this in my PR also.  Seems that there was an update of the linter.

I think its saying that the first item for any exported function should be the variable first.

ie RuntimeClassInImageCRIAPI

kiashok commented:
> > ![image](https://user-images.githubusercontent.com/99994218/279509072-fb8eaeff-4bda-4643-8e35-03f44185e247.png)
> > This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be?
> > cc @kannon92
> 
> I notice this in my PR also. Seems that there was an update of the linter.
> 
> I think its saying that the first item for any exported function should be the variable first.
> 
> ie RuntimeClassInImageCRIAPI

@kannon92 , the latest changes in the PR has already addressed this. The comment starts with the name of the featuregate. But still the linter is complaining about the same

kiashok commented:
> > > ![image](https://user-images.githubusercontent.com/99994218/279509072-fb8eaeff-4bda-4643-8e35-03f44185e247.png)
> > > This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be?
> > > cc @kannon92
> > 
> > 
> > I notice this in my PR also. Seems that there was an update of the linter.
> > I think its saying that the first item for any exported function should be the variable first.
> > ie RuntimeClassInImageCRIAPI
> 
> @kannon92 , the latest changes in the PR has already addressed this. The comment starts with the name of the featuregate. But still the linter is complaining about the same

![image](https://github.com/kubernetes/kubernetes/assets/99994218/d17b4c62-e63a-4175-a7d2-b3f2f296c797)

derekwaynecarr commented:
/approve

aravindhp commented:
> > > > ![image](https://user-images.githubusercontent.com/99994218/279509072-fb8eaeff-4bda-4643-8e35-03f44185e247.png)
> > > > This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be?
> > > > cc @kannon92
> > > 
> > > 
> > > I notice this in my PR also. Seems that there was an update of the linter.
> > > I think its saying that the first item for any exported function should be the variable first.
> > > ie RuntimeClassInImageCRIAPI
> > 
> > 
> > @kannon92 , the latest changes in the PR has already addressed this. The comment starts with the name of the featuregate. But still the linter is complaining about the same
> 

Looks like the linter is complaining about the comment starting with `// owner: @kiashok` which is strange given that is pattern elsewhere. As an experiment, I suggest starting the comment with the description and moving the rest below to see if the linter reacts differently. But this is going against the pattern in the file, so I am not sure how you should proceed.
PS: you can try different combinations locally and figure out what works.

aravindhp commented:
> > > > > This comment has already been fixed in the PR but it still complains about it. Does anyone know why this could be?
> > > > > cc @kannon92
> > > > 
> > > > 
> > > > I notice this in my PR also. Seems that there was an update of the linter.
> > > > I think its saying that the first item for any exported function should be the variable first.
> > > > ie RuntimeClassInImageCRIAPI
> > > 
> > > 
> > > @kannon92 , the latest changes in the PR has already addressed this. The comment starts with the name of the featuregate. But still the linter is complaining about the same
> 
> Looks like the linter is complaining about the comment starting with `// owner: @kiashok` which is strange given that is pattern elsewhere. As an experiment, I suggest starting the comment with the description and moving the rest below to see if the linter reacts differently. But this is going against the pattern in the file, so I am not sure how you should proceed. PS: you can try different combinations locally and figure out what works.

@kiashok Oh wait, this is with the `pull-kubernetes-linter-hints` job which is optional. I think you can ignore that and stick with pattern existing in the file.

mikebrow reviewed (approved):
/lgtm

dims commented:
`golangci-lint-pr-hints` has caught something, please fix.

kiashok commented:
> `golangci-lint-pr-hints` has caught something, please fix.

@dims This seems like a new linter check and not sure what its complaining about:

![image](https://github.com/kubernetes/kubernetes/assets/99994218/c69f8f2b-f042-415c-bd79-8dd6dad9fca7)

Any combination that I try is failing this linter. Other PRs that went in few hours back are starting their comments with // Owners just like the rest of this file. Do you happen to know what this linter check is about?

jsturtevant commented on pkg/features/kube_features.go:
```diff
@@ -710,6 +710,14 @@ const (
 	// certificate as expiration approaches.
 	RotateKubeletServerCertificate featuregate.Feature = "RotateKubeletServerCertificate"
 
+	// RuntimeClassInImageCriAPI
+	// owner: @kiashok
+	// kep: https://kep.k8s.io/4216
+	// alpha: v1.29
+	//
+	// RuntimeClassInImageCriAPI adds support to pull images based on the runtime class specified
+	RuntimeClassInImageCriAPI featuregate.Feature = "RuntimeClassInImageCriApi"
+
```
```suggestion
	// owner: @kiashok
	// kep: https://kep.k8s.io/4216
	// alpha: v1.29
	//
	// adds support to pull images based on the runtime class specified
	RuntimeClassInImageCriAPI featuregate.Feature = "RuntimeClassInImageCriApi"

```

jsturtevant replied:
would match the rest of the entries in the file

dims commented:
@kiashok if it doesn't make sense, you can use `// nolint:revive` to shut it off :) ( see https://github.com/search?type=code&q=%2Fnolint.*revive%2F+language%3AGo and https://golangci-lint.run/usage/false-positives/ )

jsturtevant commented:
/lgtm

kannon92 commented:
> > `golangci-lint-pr-hints` has caught something, please fix.
> 
> @dims This seems like a new linter check and not sure what its complaining about:
> 
> ![image](https://user-images.githubusercontent.com/99994218/279531755-c69f8f2b-f042-415c-bd79-8dd6dad9fca7.png)
> 
> Any combination that I try is failing this linter. Other PRs that went in few hours back are starting their comments with // Owners just like the rest of this file. Do you happen to know what this linter check is about?

```
	// KubeletSeparateDiskGC enables Kubelet to garbage collection images/containers on different filesystems
	// owner: @kannon92
	// kep: https://kep.k8s.io/4191
	// alpha: v1.29
	KubeletSeparateDiskGC featuregate.Feature = "KubeletSeparateDiskGC"
```

Trying this in my PR.  I don't love this change as it goes against all the other kubelet features.  But it makes sense for exported functions.

kiashok commented:
> > > `golangci-lint-pr-hints` has caught something, please fix.
> > 
> > 
> > @dims This seems like a new linter check and not sure what its complaining about:
> > ![image](https://user-images.githubusercontent.com/99994218/279531755-c69f8f2b-f042-415c-bd79-8dd6dad9fca7.png)
> > Any combination that I try is failing this linter. Other PRs that went in few hours back are starting their comments with // Owners just like the rest of this file. Do you happen to know what this linter check is about?
> 
> ```
> 	// KubeletSeparateDiskGC enables Kubelet to garbage collection images/containers on different filesystems
> 	// owner: @kannon92
> 	// kep: https://kep.k8s.io/4191
> 	// alpha: v1.29
> 	KubeletSeparateDiskGC featuregate.Feature = "KubeletSeparateDiskGC"
> ```
> 
> Trying this in my PR. I don't love this change as it goes against all the other kubelet features. But it makes sense for exported functions.

yeah! The other PRs that went in for kubelet features, seem to have ignored this and kept to the format of the rest of the file which is // Owner followed by the rest

jsturtevant commented:
I am in favor of matching the rest of the feature gates comments as it is now and letting the PR merge as is (since this isn't blocking).  

We should create a follow up issue to sort his out.  I don't think adding the `nolint:revive` makes sense here as it matches the rest of the comments in this file /cc @dims

kiashok commented:
> golangci-lint-pr-hints

Opened a new github issue to fix this https://github.com/kubernetes/kubernetes/issues/121660

sftim commented:
Changelog suggestion
```markdown
Added new feature gate `RuntimeClassInImageCriApi` to support per-RuntimeClass container image pulls

Notable changes:
1. Populate new `RuntimeHandler` field in CRI's `ImageSpec` struct during image pulls from container runtimes.
2. Pass `runtimeHandler` field in CRI `RemoveImage()` call to container runtime in kubelet's image garbage collection.
```
